File: projects/rpn2tex/migrations/rust-module-by-module-2/tests/parser_integration_test.rs
//! Integration tests for the parser module.

use rpn2tex::{Lexer, Parser};

#[test]
fn test_parser_integration_simple_addition() {
    let mut lexer = Lexer::new("5 3 +");
    let tokens = lexer.tokenize().unwrap();
    let mut parser = Parser::new(tokens);
    let ast = parser.parse().unwrap();

    // Verify the AST structure
    match ast {
        rpn2tex::Expr::BinaryOp { operator, .. } => {
            assert_eq!(operator, "+");
        }
        _ => panic!("Expected BinaryOp"),
    }
}

#[test]
fn test_parser_integration_precedence() {
    let mut lexer = Lexer::new("5 3 + 2 *");
    let tokens = lexer.tokenize().unwrap();
    let mut parser = Parser::new(tokens);
    let ast = parser.parse().unwrap();

    // Should create (5+3)*2
    match ast {
        rpn2tex::Expr::BinaryOp { operator, left, .. } => {
            assert_eq!(operator, "*");
            match *left {
                rpn2tex::Expr::BinaryOp { operator, .. } => {
                    assert_eq!(operator, "+");
                }
                _ => panic!("Expected nested BinaryOp"),
            }
        }
        _ => panic!("Expected BinaryOp"),
    }
}

#[test]
fn test_parser_integration_multiple_operations() {
    let mut lexer = Lexer::new("1 2 + 3 4 + *");
    let tokens = lexer.tokenize().unwrap();
    let mut parser = Parser::new(tokens);
    let ast = parser.parse().unwrap();

    // Should create (1+2)*(3+4)
    match ast {
        rpn2tex::Expr::BinaryOp {
            operator,
            left,
            right,
            ..
        } => {
            assert_eq!(operator, "*");
            assert!(matches!(*left, rpn2tex::Expr::BinaryOp { .. }));
            assert!(matches!(*right, rpn2tex::Expr::BinaryOp { .. }));
        }
        _ => panic!("Expected BinaryOp"),
    }
}

#[test]
fn test_parser_integration_floating_point() {
    let mut lexer = Lexer::new("3.14 2 *");
    let tokens = lexer.tokenize().unwrap();
    let mut parser = Parser::new(tokens);
    let ast = parser.parse().unwrap();

    match ast {
        rpn2tex::Expr::BinaryOp { operator, left, .. } => {
            assert_eq!(operator, "*");
            match *left {
                rpn2tex::Expr::Number { value, .. } => {
                    assert_eq!(value, "3.14");
                }
                _ => panic!("Expected Number"),
            }
        }
        _ => panic!("Expected BinaryOp"),
    }
}

#[test]
fn test_parser_integration_negative_number() {
    let mut lexer = Lexer::new("-5 3 +");
    let tokens = lexer.tokenize().unwrap();
    let mut parser = Parser::new(tokens);
    let ast = parser.parse().unwrap();

    match ast {
        rpn2tex::Expr::BinaryOp { operator, left, .. } => {
            assert_eq!(operator, "+");
            match *left {
                rpn2tex::Expr::Number { value, .. } => {
                    assert_eq!(value, "-5");
                }
                _ => panic!("Expected Number"),
            }
        }
        _ => panic!("Expected BinaryOp"),
    }
}

#[test]
fn test_parser_integration_error_insufficient_operands() {
    let mut lexer = Lexer::new("5 +");
    let tokens = lexer.tokenize().unwrap();
    let mut parser = Parser::new(tokens);
    let result = parser.parse();

    assert!(result.is_err());
    let err = result.unwrap_err();
    assert!(err.message.contains("requires two operands"));
}

#[test]
fn test_parser_integration_error_excess_operands() {
    let mut lexer = Lexer::new("5 3 2 +");
    let tokens = lexer.tokenize().unwrap();
    let mut parser = Parser::new(tokens);
    let result = parser.parse();

    assert!(result.is_err());
    let err = result.unwrap_err();
    assert!(err.message.contains("Invalid RPN"));
}

#[test]
fn test_parser_integration_error_empty_expression() {
    let mut lexer = Lexer::new("");
    let tokens = lexer.tokenize().unwrap();
    let mut parser = Parser::new(tokens);
    let result = parser.parse();

    assert!(result.is_err());
    let err = result.unwrap_err();
    assert!(err.message.contains("Empty expression"));
}

#[test]
fn test_parser_integration_all_operators() {
    let test_cases = vec![
        ("5 3 +", "+"),
        ("5 3 -", "-"),
        ("5 3 *", "*"),
        ("5 3 /", "/"),
    ];

    for (input, expected_op) in test_cases {
        let mut lexer = Lexer::new(input);
        let tokens = lexer.tokenize().unwrap();
        let mut parser = Parser::new(tokens);
        let ast = parser.parse().unwrap();

        match ast {
            rpn2tex::Expr::BinaryOp { operator, .. } => {
                assert_eq!(operator, expected_op, "Failed for input: {}", input);
            }
            _ => panic!("Expected BinaryOp for input: {}", input),
        }
    }
}

#[test]
fn test_parser_integration_complex_expression() {
    let mut lexer = Lexer::new("10 2 / 3 + 4 *");
    let tokens = lexer.tokenize().unwrap();
    let mut parser = Parser::new(tokens);
    let ast = parser.parse().unwrap();

    // Should create (10/2+3)*4
    match ast {
        rpn2tex::Expr::BinaryOp { operator, .. } => {
            assert_eq!(operator, "*");
        }
        _ => panic!("Expected BinaryOp"),
    }
}

File: projects/rpn2tex/migrations/java-module-by-module-2/src/test/java/com/rpn2tex/ParserIntegrationTest.java
package com.rpn2tex;

import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Integration tests for the {@link Parser} class working with the {@link Lexer}.
 *
 * <p>These tests validate the complete tokenization and parsing pipeline,
 * ensuring that real RPN expressions from strings are correctly converted to AST nodes.
 */
class ParserIntegrationTest {

    @Test
    void testLexerToParserSimpleAddition() throws RpnException {
        // Test the full pipeline: String -> Lexer -> Tokens -> Parser -> AST
        String input = "5 3 +";

        Lexer lexer = new Lexer(input);
        List<Token> tokens = lexer.tokenize();

        Parser parser = new Parser(tokens);
        Expr result = parser.parse();

        assertInstanceOf(BinaryOp.class, result);
        BinaryOp op = (BinaryOp) result;
        assertEquals("+", op.getOperator());
        assertEquals("5", ((Number) op.getLeft()).getValue());
        assertEquals("3", ((Number) op.getRight()).getValue());
    }

    @Test
    void testLexerToParserComplexExpression() throws RpnException {
        // "5 3 + 2 *" -> (5 + 3) * 2
        String input = "5 3 + 2 *";

        Lexer lexer = new Lexer(input);
        List<Token> tokens = lexer.tokenize();

        Parser parser = new Parser(tokens);
        Expr result = parser.parse();

        assertInstanceOf(BinaryOp.class, result);
        BinaryOp op = (BinaryOp) result;
        assertEquals("*", op.getOperator());

        // Left should be BinaryOp(+, 5, 3)
        assertInstanceOf(BinaryOp.class, op.getLeft());
        BinaryOp leftOp = (BinaryOp) op.getLeft();
        assertEquals("+", leftOp.getOperator());

        // Right should be Number(2)
        assertInstanceOf(Number.class, op.getRight());
        assertEquals("2", ((Number) op.getRight()).getValue());
    }

    @Test
    void testLexerToParserWithFloatingPoint() throws RpnException {
        String input = "3.14 2 *";

        Lexer lexer = new Lexer(input);
        List<Token> tokens = lexer.tokenize();

        Parser parser = new Parser(tokens);
        Expr result = parser.parse();

        assertInstanceOf(BinaryOp.class, result);
        BinaryOp op = (BinaryOp) result;
        assertEquals("*", op.getOperator());
        assertEquals("3.14", ((Number) op.getLeft()).getValue());
        assertEquals("2", ((Number) op.getRight()).getValue());
    }

    @Test
    void testLexerToParserNestedPrecedence() throws RpnException {
        // "2 3 4 * +" -> 2 + (3 * 4)
        String input = "2 3 4 * +";

        Lexer lexer = new Lexer(input);
        List<Token> tokens = lexer.tokenize();

        Parser parser = new Parser(tokens);
        Expr result = parser.parse();

        assertInstanceOf(BinaryOp.class, result);
        BinaryOp op = (BinaryOp) result;
        assertEquals("+", op.getOperator());

        // Left should be Number(2)
        assertInstanceOf(Number.class, op.getLeft());
        assertEquals("2", ((Number) op.getLeft()).getValue());

        // Right should be BinaryOp(*, 3, 4)
        assertInstanceOf(BinaryOp.class, op.getRight());
        BinaryOp rightOp = (BinaryOp) op.getRight();
        assertEquals("*", rightOp.getOperator());
        assertEquals("3", ((Number) rightOp.getLeft()).getValue());
        assertEquals("4", ((Number) rightOp.getRight()).getValue());
    }

    @Test
    void testLexerToParserTwoComplexSubexpressions() throws RpnException {
        // "1 2 + 3 4 + *" -> (1 + 2) * (3 + 4)
        String input = "1 2 + 3 4 + *";

        Lexer lexer = new Lexer(input);
        List<Token> tokens = lexer.tokenize();

        Parser parser = new Parser(tokens);
        Expr result = parser.parse();

        assertInstanceOf(BinaryOp.class, result);
        BinaryOp op = (BinaryOp) result;
        assertEquals("*", op.getOperator());

        // Left should be BinaryOp(+, 1, 2)
        assertInstanceOf(BinaryOp.class, op.getLeft());
        BinaryOp leftOp = (BinaryOp) op.getLeft();
        assertEquals("+", leftOp.getOperator());
        assertEquals("1", ((Number) leftOp.getLeft()).getValue());
        assertEquals("2", ((Number) leftOp.getRight()).getValue());

        // Right should be BinaryOp(+, 3, 4)
        assertInstanceOf(BinaryOp.class, op.getRight());
        BinaryOp rightOp = (BinaryOp) op.getRight();
        assertEquals("+", rightOp.getOperator());
        assertEquals("3", ((Number) rightOp.getLeft()).getValue());
        assertEquals("4", ((Number) rightOp.getRight()).getValue());
    }

    @Test
    void testLexerToParserInsufficientOperands() {
        // "5 +" should fail at parser stage
        String input = "5 +";

        Lexer lexer = new Lexer(input);
        assertDoesNotThrow(() -> {
            List<Token> tokens = lexer.tokenize();
            Parser parser = new Parser(tokens);

            RpnException exception = assertThrows(RpnException.class, parser::parse);
            assertTrue(exception.getMessage().contains("Operator '+' requires two operands"));
        });
    }

    @Test
    void testLexerToParserExtraOperands() {
        // "5 3" should fail at parser stage (missing operator)
        String input = "5 3";

        Lexer lexer = new Lexer(input);
        assertDoesNotThrow(() -> {
            List<Token> tokens = lexer.tokenize();
            Parser parser = new Parser(tokens);

            RpnException exception = assertThrows(RpnException.class, parser::parse);
            assertTrue(exception.getMessage().contains("Invalid RPN: 2 values remain on stack"));
        });
    }

    @Test
    void testLexerToParserMultipleSubtractions() throws RpnException {
        // "5 3 - 2 -" -> (5 - 3) - 2
        String input = "5 3 - 2 -";

        Lexer lexer = new Lexer(input);
        List<Token> tokens = lexer.tokenize();

        Parser parser = new Parser(tokens);
        Expr result = parser.parse();

        assertInstanceOf(BinaryOp.class, result);
        BinaryOp op = (BinaryOp) result;
        assertEquals("-", op.getOperator());

        // Left should be BinaryOp(-, 5, 3)
        assertInstanceOf(BinaryOp.class, op.getLeft());
        BinaryOp leftOp = (BinaryOp) op.getLeft();
        assertEquals("-", leftOp.getOperator());
        assertEquals("5", ((Number) leftOp.getLeft()).getValue());
        assertEquals("3", ((Number) leftOp.getRight()).getValue());

        // Right should be Number(2)
        assertInstanceOf(Number.class, op.getRight());
        assertEquals("2", ((Number) op.getRight()).getValue());
    }

    @Test
    void testLexerToParserMultipleDivisions() throws RpnException {
        // "100 10 / 5 / 2 /" -> ((100 / 10) / 5) / 2
        String input = "100 10 / 5 / 2 /";

        Lexer lexer = new Lexer(input);
        List<Token> tokens = lexer.tokenize();

        Parser parser = new Parser(tokens);
        Expr result = parser.parse();

        assertInstanceOf(BinaryOp.class, result);
        BinaryOp op = (BinaryOp) result;
        assertEquals("/", op.getOperator());

        // Verify nested structure (left-to-right associativity)
        assertInstanceOf(BinaryOp.class, op.getLeft());
        BinaryOp leftOp = (BinaryOp) op.getLeft();
        assertEquals("/", leftOp.getOperator());

        assertInstanceOf(BinaryOp.class, leftOp.getLeft());
        BinaryOp leftLeftOp = (BinaryOp) leftOp.getLeft();
        assertEquals("/", leftLeftOp.getOperator());
        assertEquals("100", ((Number) leftLeftOp.getLeft()).getValue());
        assertEquals("10", ((Number) leftLeftOp.getRight()).getValue());
    }
}
